### ISSUE TITLE
Improve live preview of layout collections variables changes

### ISSUE THREAD

**Issue by PatrykSzwer** [Starcounter/Blending#108]
Right now you can't see any reaction during making changes to collection variables unless you click "Save" button. Eventually, we want user to be able to see his changes right after writing a new value to every variable.

**Comment by kegor** [comment]
Just thoughts aloud: the initial idea can be continued to this checkbox:
![image](https://user-images.githubusercontent.com/17431807/37774086-4ed53346-2df0-11e8-8023-e6fd216a3b59.png)

**Comment by kegor** [comment]
Theoretically, we can achieve this by following steps:
1. In BE CollectionPage handler we can get Collection/CssVariable names and changed value.
2. Then using the Blender Mapping we can pass these parameters into the BP handler.
3. In the separate method, we can recalculate CssVariables string and replace the old value in each LayoutPage.CssVariables which can be found in current session.
Probably the main side effect of this approach is that if a user leaves tab after changing CSS variable value without saving then all current changes are left on the page. And seems like we need yet another handler to revert changes back.

So maybe make sense to try to implement this feature on a client-side.

**Comment by kegor** [comment]
@warpech, @un-tone what do you think?

**Comment by un-tone** [comment]
At first glance, it seemed over-complicated to me if using habitual approaches here, I mean all work on server side, especially considering the requirement to revert changes back if a user goes away, cancels changes or maybe does something else.
But probably implementing on a client side will not be easier because it will also require implementing reverting the CSS styles, it is unclear to me how it can be done because the client side should be in sync with the server side.

Probably, it is possible to make some changes in and improve the CSS Variables implementation to make it more convenient to use and manipulate with its data. Currently, it looks quite hard to do so. Just my point of view, I could miss something.

**Comment by warpech** [comment]
Regarding the problem in the OR, it seems to me that we already have the server and client solution, just for some reason the transaction is isolated until the `Save` button is pressed on the CSS variables editing page.

The two-way data binding between the view-model and the view is already defined here: https://github.com/Starcounter/Blending/blob/4486d941cd5ed6ffc6c83fa69b456eb4e13cef66/src/BlendingProvider/Website/Api/ContentHandlers.cs#L75

The view-model uses the getter: https://github.com/Starcounter/Blending/blob/4486d941cd5ed6ffc6c83fa69b456eb4e13cef66/src/BlendingProvider/Website/ViewModels/LayoutPage.json.cs#L15. This getter re-executes each time a change comes from the client.

---

Regarding @kegor's comment illustrated by the screenshot, the code in https://github.com/Starcounter/Blending/blob/4486d941cd5ed6ffc6c83fa69b456eb4e13cef66/src/BlendingProvider/Website/Api/ContentHandlers.cs#L75 should already cater for that. Doesn't it?

**Comment by kegor** [comment]
Hi @warpech, for me it seems like the code that you mentioned works as expected only on page refresh, which called after clicking `Save` by this line:
https://github.com/Starcounter/Blending/blob/4486d941cd5ed6ffc6c83fa69b456eb4e13cef66/src/BlendingEditor/CssVariablesManager/ViewModels/CollectionPage.json.cs#L35

@PatrykSzwer maybe you have an idea: can we use it right after losing focus while editing the CssVariable value?

**Comment by PatrykSzwer** [comment]
If the situation is how you described, I cannot think of other solution immediately than just doing a little cheat, like on loosing focus changing some part of the view-model trough script and then catching this change on server side trough Handle and refreshing, but it seems like really inefficient way. At least it can be done to make sure that is the problem.

**Comment by PatrykSzwer** [comment]
I was doing some external investigation about causes of this issue and here are some informations:
Collection page, which is responsible for changing collection css variables values is part of the BlendingEditor app. Layout page, which is used to provide layouts for pages, including applied css collections with variables is part of the BlendingProvider app. BlendingEditor and BlendingProvider are two separate apps that have also separated scopes of transactions. In layout page, there is a specific part that provides css collections from database:
https://github.com/Starcounter/Blending/blob/master/src/BlendingProvider/Website/ViewModels/LayoutPage.json.cs#L20

after small debug session you could see that CssVariables property get is firing up in every situation, even during changes of CssVariables in BlendingEditor *without commiting transaction*, which is good sign. Although without calling `Transaction.Commit()` in BE values of css variables that BP got there are not updated. This is due to fact that those two things, changed variables by blendingEditor and this get in BlendingProvider are in different scopes of transactions.

In conclusion the only way to achieve this feature is to somehow use the same transaction scope in both those situations. I talked to @chrhol about possible solution, but he wasn't really familiar with a way how scope is set in BlendingEditor (Authorization Library is used for that: https://github.com/Starcounter/Blending/blob/master/src/BlendingEditor/Website/Api/Authorization/AuthorizedHandlers.cs#L69 )
I guess the next possible step to do is to ask @joozek78 how he sees eventual possibility of achieving same scopes in both apps with authorization library in the middle.

**Comment by warpech** [comment]
In this issue we experience some problem with long running transactions. @kegor will try to make a minimum app that exhibits the problem.

**Comment by kegor** [comment]
Hi @warpech, here I pushed very simple and very default app with two pages: https://github.com/kegor/SmallApps/commit/38117ccdd800364debea281add46c331c7a1621e, is this the case that interests us to start an investigation?

![long-transaction-two-pages](https://user-images.githubusercontent.com/17431807/42173225-04779b40-7e27-11e8-8ceb-40e56d552994.gif)

**Comment by kegor** [comment]
Similar case but now Pages are storing in the Session.Store:
![long-transaction-two-pages_with_store](https://user-images.githubusercontent.com/17431807/42217108-9dc9cd68-7ecc-11e8-9ebc-7f1266659075.gif)

**Comment by kegor** [comment]
![long-transaction-two-pages_success](https://user-images.githubusercontent.com/17431807/42228384-82ce723c-7eec-11e8-8ba6-3165821cc05e.gif)
And this looks like what we actually need (the related commit: https://github.com/kegor/SmallApps/commit/955098715fd78c43b7926fe23ab0d4b60433587b), I just used the approach mentioned in the docs https://docs.starcounter.io/v/2.4/topic-guides/transactions/long-running-transactions#attaching-a-transaction-to-an-existing-json-object:
1. Get the first page which `AttachedScope` we are interesting.
2. Use this `FirstPage.AttachedScope` for scoping the construction of the `SecondPage`.

@warpech if you agree with this approach, I plan to try same steps but now with two apps.

**Comment by warpech** [comment]
I made another view in a fork of this app to demonstrate how to achieve a shared transaction. It appears to me that this is possible only if two view-models are attached to a JSON tree that is the current main JSON response. 

Screen recording:

![output](https://user-images.githubusercontent.com/566463/42230161-9129dbd6-7ee8-11e8-99a8-9551b69fd1c6.gif)

Source: https://github.com/warpech/SmallApps/commit/e1a5a4d4a931740a91dee07767f3227c33d8c45c

**Comment by warpech** [comment]
> @warpech if you agree with this approach, I plan to try same steps but now with two apps.

That's a great finding, I'm happy that we now understand long-running transactions better.

But the docs show a simplistic case in which all code is in one app. In case of Blending app suite, CSS variables rendering comes from BlendingProvider and "CSS collection page" comes from BlendingEditor. 

How can you "Get the first page which has `AttachedScope` interesting for us" if that page is in a different app? Maybe through blending a technical view-model from BlendingProvider to BlendingEditor's "CSS collection page"?

**Comment by kegor** [comment]
@warpech, I tried to add the second app and using `Blender.MapUri` here: https://github.com/kegor/SmallApps/commit/6bb8621c12d7b5c3ce6cb614608e3e973496dce4 (I used a direct reference from App2 to App1 to use same DB table `Person` for me it looks like using `Simplified` and I think this is not affecting the main case): and now I see some positive results:
![long-transaction-two-apps_with_mapuri_success](https://user-images.githubusercontent.com/17431807/42249343-6e0940d8-7f33-11e8-837a-67133aec9ab5.gif)
The conclusions to the moment:
1. Seems like using `Blender.MapUri` really helped us to share the transaction scope between two different pages of two different apps, most likely because second view-model was attached into JSON tree.
2. Using shared data from common long transactions starts working for me after commenting explicit `Db.Scope` wrapping in the second app: https://github.com/kegor/SmallApps/commit/6bb8621c12d7b5c3ce6cb614608e3e973496dce4#diff-88e711715cc697b8e2f885e17636ff33R35
3. The expected case working only in one browser tab, most likely because of different sessions in each tab.

@warpech feel free pls to correct my conclusions. And I'll continue with thinking: can we use these findings in case of the task for CSS Collections.
